xmtp_mls/src/storage/sql_key_store.rs

```diff
@@ -0,0 +1,118 @@
+use log::{debug, error};
+use openmls_traits::key_store::{MlsEntity, OpenMlsKeyStore};
+
+use crate::{Delete, Fetch, Store};
+
+use super::{
+    encrypted_store::models::StoredKeyStoreEntry,
+    serialization::{db_deserialize, db_serialize},
+    EncryptedMessageStore, StorageError,
+};
+
+#[derive(Debug)]
+pub struct SqlKeyStore<'a> {
+    store: &'a EncryptedMessageStore,
+}
+
+impl Default for SqlKeyStore<'_> {
+    fn default() -> Self {
+        unimplemented!()
+    }
+}
+
+impl<'a> SqlKeyStore<'a> {
+    pub fn new(store: &'a EncryptedMessageStore) -> Self {
+        SqlKeyStore { store }
+    }
+}
+
+impl OpenMlsKeyStore for SqlKeyStore<'_> {
+    /// The error type returned by the [`OpenMlsKeyStore`].
+    type Error = StorageError;
+
+    /// Store a value `v` that implements the [`ToKeyStoreValue`] trait for
+    /// serialization for ID `k`.
+    ///
+    /// Returns an error if storing fails.
+    fn store<V: MlsEntity>(&self, k: &[u8], v: &V) -> Result<(), Self::Error> {
+        let entry = StoredKeyStoreEntry {
+            key_bytes: k.to_vec(),
+            value_bytes: db_serialize(v)?,
+        };
+        entry.store(&mut self.store.conn()?)?;
+        Ok(())
+    }
+
+    /// Read and return a value stored for ID `k` that implements the
+    /// [`FromKeyStoreValue`] trait for deserialization.
+    ///
+    /// Returns [`None`] if no value is stored for `k` or reading fails.
+    fn read<V: MlsEntity>(&self, k: &[u8]) -> Option<V> {
+        let conn_result = self.store.conn();
```

Comment:
This is going to get a little tricky when we need to interact with OpenMLS inside a transaction.

We don't control the interface for `OpenMlsKeyStore`, so we can't add an argument for a connection. I guess we could generate new `OpenMlsKeyStore` instances that hold a `conn` as a property if we needed a version that was transaction-friendly.

Reply:
Good point - do you have an example of where we might need to do that? That might help with thinking through this. I've been hoping that we can treat OpenMLS data as separate and opaque to us - they handle their stuff, we handle ours, but aware that this abstraction can break

Reply:
I guess we can wait and see, if it's not immediately obvious

Reply:
The one I'm thinking about is processing an incoming message from the group topic. Ideally we would wrap each message processing task into a transaction, and roll the whole thing back if something went wrong. But some of the messages are going to need writes to the Keystore as group state gets updated.